docs/reference/content/driver/reference/gridfs/index.md

```diff
@@ -0,0 +1,201 @@
+
+ObjectId fileId = gridFSBucket.uploadFromStream("mongodb-tutorial", streamToUploadFrom, options);
+```
+
+### OpenUploadStream
+
+The [`openUploadStream`]({{< apiref "com/mongodb/client/gridfs/GridFSBucket.html#openUploadStream-java.lang.String-com.mongodb.client.gridfs.model.GridFSUploadOptions-">}}) method returns a [`GridFSUploadStream`]({{< apiref "mongodb/client/gridfs/GridFSUploadStream.html">}}) which extends [`OutputStream`](http://docs.oracle.com/javase/8/docs/api/index.html?java/io/OutputStream.html) and can be written to.
+
+The `GridFSUploadStream` buffers data until it reaches the `chunkSizeBytes` and then inserts the chunk into the chunks collection.  When the `GridFSUploadStream` is closed, the final chunk is written and the file metadata is inserted into the files collection.
+
+The following example uploads an into `GridFSBucket` via the returned `OutputStream`:
+
+```java
+byte[] data = "Data to upload into GridFS".getBytes(StandardCharsets.UTF_8);
+GridFSUploadStream uploadStream = gridFSBucket.openUploadStream("sampleData", options);
+uploadStream.write(data);
+uploadStream.close();
+System.out.println("The fileId of the uploaded file is: " + uploadStream.getFileId().toHexString());
+```
+
+{{% note %}}
+GridFS will automatically create indexes on the files and chunks collections on first upload of data into the GridFS bucket.
+{{% /note %}}
+
+## Finding files stored in GridFS
+
+To find what files are listed in the `GridFSBucket` use the [`find`]({{< apiref "com/mongodb/client/gridfs/GridFSBucket.html#find--">}}) method.
+
+The following example prints out the filename of each file stored:
+
+```java
+gridFSBucket.find().forEach(
+  new Block<GridFSFile>() {
+    @Override
+    public void apply(final GridFSFile gridFSFile) {
+        System.out.println(gridFSFile.getFilename());
+    }
+});
+```
+
+You can also provide a custom filter to limit the results returned. The following example prints out the filenames of all files with a "image/png" value set as the contentType in the user defined metadata document:
+
+```java
+gridFSBucket.find(eq("metadata.contentType", "image/png")).forEach(
+  new Block<GridFSFile>() {
+      @Override
+      public void apply(final GridFSFile gridFSFile) {
+          System.out.println(gridFSFile.getFilename());
+      }
+  });
+```
+
+## Downloading from GridFS
+
+There are four main ways to download data from GridFS.
+
+### DownloadFromStream
+
+The [`downloadToStream`]({{< apiref "com/mongodb/client/gridfs/GridFSBucket.html#downloadToStream-org.bson.types.ObjectId-java.io.OutputStream-" >}}) method reads the contents from MongoDB and writes the data directly to the provided [`OutputStream`](http://docs.oracle.com/javase/8/docs/api/index.html?java/io/OutputStream.html).
+
+The following example downloads a file into the provided `OutputStream`:
+
+```java
+ByteArrayOutputStream streamToDownloadTo = new ByteArrayOutputStream();
```

Comment:
Just so people don't get the wrong idea and blow out their memory, let's use FileOutputStream instead of ByteArrayOutputStream.